rust/otap-dataflow/crates/otlp/src/fake_signal_receiver/receiver.rs

```diff
@@ -0,0 +1,308 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! Implementation of the OTLP receiver node
+//!
+//! ToDo: implement Ack and Nack control message, wait for receiver node to receive a Ack control message then the service can send a response back
+//! ToDo: Implement proper deadline function for Shutdown ctrl msg
+//!
+
+use crate::FAKE_SIGNAL_RECEIVERS;
+
+use crate::fake_signal_receiver::config::{Config, OTLPSignal, ScenarioStep, SignalConfig};
+use async_trait::async_trait;
+use linkme::distributed_slice;
+use otap_df_engine::error::Error;
+use otap_df_engine::message::ControlMsg;
+use otap_df_engine::shared::{SharedReceiverFactory, receiver as shared};
+use serde_json::Value;
+use tokio::time::{Duration, sleep};
+
+/// A Receiver that listens for OTLP messages
+pub struct FakeSignalReceiver {
+    config: Config,
+}
+
+/// Declares the Fake Signal receiver as a shared receiver factory
+///
+/// Unsafe code is temporarily used here to allow the use of `distributed_slice` macro
+/// This macro is part of the `linkme` crate which is considered safe and well maintained.
+#[allow(unsafe_code)]
+#[distributed_slice(FAKE_SIGNAL_RECEIVERS)]
+pub static FAKE_SIGNAL_RECEIVER: SharedReceiverFactory<OTLPSignal> = SharedReceiverFactory {
+    name: "urn:otel:fake:signal:receiver",
+    create: |config: &Value| Box::new(FakeSignalReceiver::from_config(config)),
+};
+
+impl FakeSignalReceiver {
+    /// creates a new FakeSignalReceiver
+    #[must_use]
+    pub fn new(config: Config) -> Self {
+        FakeSignalReceiver { config }
+    }
+
+    /// Creates a new FakeSignalReceiver from a configuration object
+    #[must_use]
+    pub fn from_config(config: &Value) -> Self {
+        let config: Config =
+            serde_json::from_value(config.clone()).unwrap_or_else(|_| Config::new(vec![]));
+        FakeSignalReceiver { config }
+    }
+}
+
+// Use the async_trait due to the need for thread safety because of tonic requiring Send and Sync traits
+// The Shared version of the receiver allows us to implement a Receiver that requires the effect handler to be Send and Sync
+#[async_trait]
+impl shared::Receiver<OTLPSignal> for FakeSignalReceiver {
```

Comment:
Why a shared and not a local Receiver?
We are not using Tonic in this receiver, right?